src/frontend/flight/sync.rs

```diff
@@ -0,0 +1,620 @@
+use arrow::array::RecordBatch;
+use arrow_schema::SchemaRef;
+use datafusion::datasource::TableProvider;
+use datafusion::physical_expr::expressions::{col, lit};
+use datafusion::physical_plan::filter::FilterExec;
+use datafusion::physical_plan::memory::MemoryExec;
+use datafusion::physical_plan::projection::ProjectionExec;
+use datafusion::physical_plan::union::UnionExec;
+use datafusion::physical_plan::ExecutionPlan;
+use datafusion_common::{Result, ScalarValue};
+use deltalake::kernel::{Action, Schema};
+use deltalake::logstore::LogStore;
+use deltalake::operations::create::CreateBuilder;
+use deltalake::protocol::{DeltaOperation, SaveMode};
+use deltalake::DeltaTable;
+use indexmap::IndexMap;
+use itertools::Itertools;
+use serde_json::Value;
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
+use tracing::{debug, info};
+
+use crate::context::delta::plan_to_object_store;
+
+use crate::context::SeafowlContext;
+use crate::frontend::flight::handler::SEAFOWL_SYNC_DATA_SEQUENCE_NUMBER;
+use crate::frontend::flight::SEAFOWL_SYNC_DATA_UD_FLAG;
+
+type SequenceNumber = u64;
+
+// A handler for caching, coalescing and flushing table syncs received via
+// the Arrow Flight `do_put` calls.
+//
+// Each `DataSyncCommand` that accompanies a record batch carries information on
+// the origin of change, sequence number, primary keys and whether this is the
+// last message in the sequence.
+//
+// It uses a greedy table-based algorithm for flushing: once the criteria is met
+// it will go through table's ordered by the oldest sync flushing all pending syncs
+// for that table. Special care is taken about deducing what is the correct durable
+// sequence to report back to the caller.
+//
+//           │sync #1│sync #2│sync #3│sync #4│sync #5│
```

Comment:
sync here refers to the result of an RPC? a batch basically?

Reply:
Yes, sync corresponds to a single RPC call (so a vector of batches with the same schema).